### PR TITLE
MetaClient Minor Edge Cases and Added Exception 

### PR DESCRIPTION
**Pull request by Marcosrico**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

https://github.com/apache/helix/issues/2237

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Minor changes to catch exception in edge cases. Added extra exception in Util (for puppy stress testing in future PR's)

### Tests

- [x] The following tests are written for this issue:

N/A

- The following is the result of the "mvn test" command on the appropriate module:

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/util/ZkMetaClientUtil.java [hunk 1]
```diff
@@ -235,6 +237,8 @@ public static MetaClientException translateZkExceptionToMetaclientException(ZkEx
       return new MetaClientTimeoutException(e);
     } else if (e instanceof ZkInterruptedException) {
       return new MetaClientInterruptException(e);
+    } else if (e instanceof ZkNodeExistsException) {
```
Comment (xyuanlu) [inline meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/util/ZkMetaClientUtil.java:240]: Where is this exception used?

File: meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java [hunk 2]
```diff
@@ -75,9 +76,12 @@ public DistributedSemaphore(MetaClientInterface<DataRecord> client) {
     _metaClient = client;
     try {
       _metaClient.connect();
-      // TODO: Differentiate exception catch between already connected and already closed.
     } catch (IllegalStateException e) {
-      // Ignore as it either has already been connected or already been closed.
+      if (e.getMessage().contains("connect() has already been called")) {
+        LOG.info("Client already connected");
```
Comment (xyuanlu) [inline meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java:81]: Checking test message is a hacky way. Any other options?

File: meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/util/ZkMetaClientUtil.java (hunk 1 above)
Comment (Marcosrico) [reply meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/util/ZkMetaClientUtil.java:240]: For puppy stress testing in future PR's (note in PR description)

File: meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java (hunk 2 above)
Comment (Marcosrico) [reply meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java:81]: Right but there is no other way as far as I know. There is a method that checks whether a node is closed but that is zk specific and shouldn't be added in a generic distributedSemaphore client creation. Zk.connect() throws the same IllegalStateException for both client exceptions so there isn't a way of separating the two. Furthermore, while kinda of hacky, the current implementation of zkClient.connect() exceptions is very mature and the chances of one modifying it is slim. However, this begs the question for future use cases besides zk.

Comment (xyuanlu) [reply meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java:81]: How about we add an API `isClosed` in metaclientInterface. So that we can use this generic API in recipes client.

Comment (qqu0127) [reply meta-client/src/main/java/org/apache/helix/metaclient/recipes/lock/DistributedSemaphore.java:81]: +1 to xyuanlu's point
Alternatively, you can make `_metaClient.connect()` idempotent, all following calls after initial connect() are no-op. Users won't have to worry about checking isConnected. This might lead to some behavior changes though.

File: meta-client/src/main/java/org/apache/helix/metaclient/exception/MetaClientNodeExistsException.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    super(message, cause);
+  }
+
+  public MetaClientNodeExistsException(String message) {
+    super(message);
+  }
+
+  public MetaClientNodeExistsException(Throwable cause) {
+    super(cause);
+  }
```
[the remark this answers is not in the document]
Comment (Marcosrico) [reply meta-client/src/main/java/org/apache/helix/metaclient/exception/MetaClientNodeExistsException.java:37]: These are for the Puppy stress tests!

**Review by xyuanlu — APPROVED** [review]

**Comment by Marcosrico** [comment]
PR Approved by @xyuanlu 
Commit Message: MetaClient LockClient and Semaphore Creation Edge Cases and Added Exception
